bigtable/benchmarks/setup.cc

```diff
@@ -0,0 +1,110 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "bigtable/benchmarks/setup.h"
+#include <absl/strings/str_cat.h>
+#include <cctype>
+#include <iomanip>
+#include <sstream>
+#include "bigtable/benchmarks/random.h"
+#include "bigtable/client/build_info.h"
+
+/// Supporting types and functions to implement `BenchmarkSetup`
+namespace {
+std::string FormattedStartTime() {
+  auto start = std::chrono::system_clock::now();
+  std::time_t start_c = std::chrono::system_clock::to_time_t(start);
+  std::string formatted("YYYY-MM-DDTHH:SS:MMZ");
+  std::strftime(&formatted[0], formatted.size(), "%FT%TZ",
+                std::gmtime(&start_c));
+  return formatted;
+}
+
+std::string FormattedAnnotations() {
+  std::string notes =
+      absl::StrCat(bigtable::version_string(), ";", bigtable::compiler, ";",
+                   bigtable::compiler_flags);
+  std::transform(notes.begin(), notes.end(), notes.begin(),
+                 [](char c) { return c == '\n' ? ';' : c; });
+  return notes;
+}
+
+std::string MakeRandomTableId(std::string const& prefix) {
+  static std::string const table_id_chars(
+      "ABCDEFGHIJLKMNOPQRSTUVWXYZabcdefghijlkmnopqrstuvwxyz0123456789_");
+  auto gen = bigtable::benchmarks::MakeDefaultPRNG();
+  return absl::StrCat(
+      prefix, "-",
+      bigtable::benchmarks::Sample(
+          gen, bigtable::benchmarks::kTableIdRandomLetters, table_id_chars));
+}
+}  // anonymous namespace
+
+namespace bigtable {
+namespace benchmarks {
+BenchmarkSetup::BenchmarkSetup(std::string const& prefix, int& argc,
+                               char* argv[])
+    : start_time_(FormattedStartTime()),
+      notes_(FormattedAnnotations()),
+      project_id_(),
+      instance_id_(),
+      table_id_(MakeRandomTableId(prefix)) {
+  if (argc < 3) {
+    std::string const cmd = argv[0];
+    auto last_slash = std::string(argv[0]).find_last_of('/');
+    std::cerr << "Usage: " << cmd.substr(last_slash + 1)
+              << " <project> <instance>"
+              << " [thread-count (" << kDefaultThreads << ")]"
+              << " [test-duration-seconds (" << kDefaultTestDuration << "min)]"
+              << " [table-size (" << kDefaultTableSize << ")]"
+              << " [use-embedded-server (false)]" << std::endl;
+    throw std::runtime_error("too few arguments for program.");
+  }
+
+  auto shift = [&argc, &argv]() {
+    char* r = argv[1];
+    std::copy(argv + 2, argv + argc, argv + 1);
+    --argc;
+    return r;
+  };
+
+  project_id_ = shift();
+  instance_id_ = shift();
+
+  if (argc == 1) {
+    return;
+  }
+  thread_count_ = std::stoi(shift());
```

Comment:
Does this mean you have to provide the arguments in the given order and can't provide, say, just the table-size?

Reply:
Yes it does.  And yes, I am being lazy.  I can change this to parse `--foo` arguments, but would need to introduce `gflags` as a dependency, or use `gpr_command_line` (which actually I do not know if it is exported from gRPC), or create a custom parser because C++ does not have a standard library for command-line parsing.  If you think it is terrible as is, I would be happy to change to use `gflags`.

Reply:
Meh, let's live with this until it becomes more annoying than changing it would be ;)